common/src/main/java/com/wynntils/services/itemfilter/SearchQuery.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.services.itemfilter;
+
+import com.wynntils.core.components.Services;
+import com.wynntils.models.items.WynnItem;
+import com.wynntils.utils.type.ErrorOr;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.IntStream;
+
+public class SearchQuery {
+    private final String queryString;
+
+    private final List<ItemFilter> itemFilters;
+
+    private final List<String> plainTextTokens;
+
+    private final List<Integer> ignoredCharIndices;
+
+    private final List<Integer> validFilterCharIndices;
+
+    private final List<String> errors;
+
+    protected SearchQuery(
+            String queryString,
+            List<ItemFilter> itemFilters,
+            List<Integer> ignoredCharIndices,
+            List<Integer> validFilterCharIndices,
+            List<String> errors,
+            List<String> plainTextTokens) {
+        this.queryString = queryString;
+        this.itemFilters = itemFilters;
+        this.ignoredCharIndices = ignoredCharIndices;
+        this.validFilterCharIndices = validFilterCharIndices;
+        this.errors = errors;
+        this.plainTextTokens = plainTextTokens;
+    }
+
+    public static SearchQuery fromQueryString(String queryString) {
```

Comment:
I would recommend moving this functionality to the ItemFilterService. If you do, it will be clear that this is a very thin data carrying class. In fact, it could probably be turned into a record instead of a class if you do that.

The matcher functions do also seem a bit out of place, as bolyai suggests. I'm not sure where they are used, but they could perhaps also move to the service.

Reply:
Also, if you make it a record, please move it to the `type` subpackage. (We use it for interfaces, enums and records)

Reply:
Done